File: js/packages/types/src/utilityClosetPuzzle.ts
export interface UtilityClosetPuzzle {
  id: string;
  createdAt: string;
  updatedAt: string;
  button0HueLightId?: number | null;
  button1HueLightId?: number | null;
  button2HueLightId?: number | null;
  button3HueLightId?: number | null;
  button4HueLightId?: number | null;
  button5HueLightId?: number | null;
  button0CurrentHue: number;
  button1CurrentHue: number;
  button2CurrentHue: number;
  button3CurrentHue: number;
  button4CurrentHue: number;
  button5CurrentHue: number;
  button0BaseHue: number;
  button1BaseHue: number;
  button2BaseHue: number;
  button3BaseHue: number;
  button4BaseHue: number;
  button5BaseHue: number;
  allGreensAchieved: boolean;
  allPurplesAchieved: boolean;
}

export interface ButtonConfig {
  slot: number;
  hueLightId?: number | null;
  baseHue: number;
}

export type PuzzleColor = 'Off' | 'Blue' | 'Red' | 'White' | 'Yellow' | 'Purple' | 'Green';

export const PUZZLE_COLORS: PuzzleColor[] = ['Off', 'Blue', 'Red', 'White', 'Yellow', 'Purple', 'Green'];

export const COLOR_TO_INDEX: Record<PuzzleColor, number> = {
  'Off': 0,
  'Blue': 1,
  'Red': 2,
  'White': 3,
  'Yellow': 4,
  'Purple': 5,
  'Green': 6,
};

export const INDEX_TO_COLOR: Record<number, PuzzleColor> = {
  0: 'Off',
  1: 'Blue',
  2: 'Red',
  3: 'White',
  4: 'Yellow',
  5: 'Purple',
  6: 'Green',
};

File: go/pkg/db/utility_closet_puzzle.go
package db

import (
	"context"
	"time"

	"github.com/MaxBlaushild/poltergeist/pkg/models"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type utilityClosetPuzzleHandler struct {
	db *gorm.DB
}

func (h *utilityClosetPuzzleHandler) GetPuzzle(ctx context.Context) (*models.UtilityClosetPuzzle, error) {
	var puzzle models.UtilityClosetPuzzle

	// Try to find existing puzzle
	err := h.db.WithContext(ctx).First(&puzzle).Error
	if err == gorm.ErrRecordNotFound {
		// Create new puzzle with default state if it doesn't exist
		puzzle = models.UtilityClosetPuzzle{
			ID:                uuid.New(),
			CreatedAt:         time.Now(),
			UpdatedAt:         time.Now(),
			Button0CurrentHue: 0,
			Button1CurrentHue: 1,
			Button2CurrentHue: 2,
			Button3CurrentHue: 3,
			Button4CurrentHue: 4,
			Button5CurrentHue: 5,
			Button0BaseHue:    0,
			Button1BaseHue:    1,
			Button2BaseHue:    2,
			Button3BaseHue:    3,
			Button4BaseHue:    4,
			Button5BaseHue:    5,
			AllGreensAchieved: false,
		}
		if err := h.db.WithContext(ctx).Create(&puzzle).Error; err != nil {
			return nil, err
		}
		return &puzzle, nil
	}
	if err != nil {
		return nil, err
	}

	return &puzzle, nil
}

func (h *utilityClosetPuzzleHandler) UpdatePuzzle(ctx context.Context, puzzle *models.UtilityClosetPuzzle) error {
	puzzle.UpdatedAt = time.Now()
	return h.db.WithContext(ctx).Save(puzzle).Error
}

func (h *utilityClosetPuzzleHandler) ResetPuzzle(ctx context.Context) (*models.UtilityClosetPuzzle, error) {
	puzzle, err := h.GetPuzzle(ctx)
	if err != nil {
		return nil, err
	}

	// Reset all buttons to base state
	puzzle.ResetToBaseState()

	// Update in database
	if err := h.UpdatePuzzle(ctx, puzzle); err != nil {
		return nil, err
	}

	return puzzle, nil
}

func (h *utilityClosetPuzzleHandler) DeletePuzzle(ctx context.Context, id uuid.UUID) error {
	return h.db.WithContext(ctx).Delete(&models.UtilityClosetPuzzle{}, id).Error
}
